src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp

```diff
@@ -0,0 +1,659 @@
+{
+    log_msg(LOG_DEBUG, "Inside send_s1_rel_cmd_to_target_enb\n");
+
+    UEContext *ue_ctxt = static_cast<UEContext*>(cb.getPermDataBlock());
+    if (ue_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "send_s1_rel_cmd_to_target_enb: ue context is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    S1HandoverProcedureContext *ho_ctxt =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (ho_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "send_s1_rel_cmd_to_target_enb: procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+   
+    struct s1relcmd_info s1relcmd;
+    s1relcmd.msg_type = s1_release_command;
+    s1relcmd.ue_idx = ue_ctxt->getContextID();
+    s1relcmd.enb_s1ap_ue_id = ho_ctxt->getTargetS1apEnbUeId();
+    s1relcmd.cause = ho_ctxt->getS1HoCause().s1apCause_m;
+    s1relcmd.enb_fd = ho_ctxt->getTargetEnbContextId();
+
+    // Fire and forget s1 release to target enb
+
+    /*Send message to S1AP-APP*/
+    cmn::ipc::IpcAddress destAddr =
+    { TipcServiceInstance::s1apAppInstanceNum_c };
+    MmeIpcInterface &mmeIpcIf = static_cast<MmeIpcInterface&>(compDb.getComponent(MmeIpcInterfaceCompId));   
+    mmeIpcIf.dispatchIpcMsg((char *) &s1relcmd, sizeof(s1relcmd), destAddr);
+
+    ProcedureStats::num_of_s1_rel_cmd_sent++;
+    log_msg(LOG_DEBUG, "Leaving send_s1_rel_cmd_to_target_enb \n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+ * Action handler : process_ho_cancel_req
+ ***************************************/
+ActStatus ActionHandlers::process_ho_cancel_req(ControlBlock &cb)
+{
+    S1HandoverProcedureContext *ho_ctxt =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (ho_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG, "process_ho_cancel_req: procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    MsgBuffer *msgBuf = static_cast<MsgBuffer*>(cb.getMsgData());
+    if (msgBuf == NULL)
+    {
+        log_msg(LOG_ERROR, "Failed to retrieve message buffer \n");
+        return ActStatus::HALT;
+    }
+    const s1_incoming_msg_data_t *msgData_p =
+            static_cast<const s1_incoming_msg_data_t*>(msgBuf->getDataPointer());
+    if (msgData_p == NULL)
+    {
+        log_msg(LOG_ERROR, "Failed to retrieve data buffer \n");
+        return ActStatus::HALT;
+    }
+
+    const struct handover_cancel_Q_msg &ho_cancel =
+            (msgData_p->msg_data.handover_cancel_Q_msg_m);
+    ho_ctxt->setS1HoCause(S1apCause(ho_cancel.cause));
+
+    ProcedureStats::num_of_ho_cancel_received++;
+
+    log_msg(LOG_INFO, "Leaving process_ho_cancel_req\n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+ * Action handler : send_ho_cancel_ack_to_src_enb
+ ***************************************/
+ActStatus ActionHandlers::send_ho_cancel_ack_to_src_enb(ControlBlock &cb)
+{
```

Comment:
What do we do with the UE context if source eNB says cancel handover... Continue with the same context ?

Reply:
Yes.